### PR TITLE
implement stop training when loss is nan hook

### PR DESCRIPTION
**Pull request by morkovka1337**
Implemented a hook that controls the loss value, and if it becomes NaN, stops training. Training is stopped without further validation, since there is no sense in validation when loss is NaN. Cancelling validation via removing eval hook looks a bit dirty, though.

Example of usage in the config:
```
custom_hooks = [
    dict(type='StopLossNanTrainingHook')
]
```
Note, that in order to use this hook, `runner` should be either `EpochRunnerWithCancel` or `IterBasedRunnerWithCancel`.

### CODE REVIEW

**Comment by nervana-ff** [comment]
Can one of the admins verify this patch?

**Comment by druzhkov-paul** [comment]
A few questions before we go further with the review:

1. What will happen if you don't remove the validation hook? Is validation supposed to crash in that case? Or you just introducing an optimization?
2. Are there any known real cases when training produces NaN values given a default setup for hyperparameters?
3. Do you have any plans to use this hook or just introducing just in case?

**Comment by morkovka1337** [comment]
> A few questions before we go further with the review:
> 
> 1. What will happen if you don't remove the validation hook? Is validation supposed to crash in that case? Or you just introducing an optimization?
> 2. Are there any known real cases when training produces NaN values given a default setup for hyperparameters?
> 3. Do you have any plans to use this hook or just introducing just in case?

Sure.

1. In my understanding, validation has no sense when training loss became NaN, so this is rather an optimization of the training time.
2. I don't think it could be caused by default setup. It is rather about experiment settings.
3. I do not plan use this hook personally, however, we have a corresponding [request](https://github.com/openvinotoolkit/training_extensions/issues/907) in our OTE issues. I think, it could be useful, when performing multiple experiments (e.g. different learning rates with the same model) via the command line. In this case, if an experiment is failed, next experiment can be started immediately without human interaction.

**Comment by druzhkov-paul** [comment]
I see the rationale behind this, but I'm not quite happy with validation hooks magic.

@eugene123tw @GalyaZalesskaya @Ilya-Krylov Do you find this functionality useful? Any plans to use it?

**Comment by eugene123tw** [comment]
@druzhkov-paul I think it'd be helpful for model debugging as NaN loss often suggests optimization on the model parameters or data annotation incorrectness (e.g. clipping gradients, lowering LR, inconsistency annotations, etc). That said, we should probably improve the logging in this PR since there's little info about what's happened if training is cancelled. 

While on the subject, I think we should also flag the training as failed and notified the user if possible.

File: mmdet/apis/ote/extension/utils/runner.py [hunk 1]
```diff
@@ -61,12 +62,12 @@ def train(self, data_loader, **kwargs):
             if self.stop():
                 break
             self._iter += 1
-
         self.call_hook('after_train_epoch')
         self.stop()
         self._epoch += 1
 
 
+
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/extension/utils/runner.py]:
```suggestion
```

File: mmdet/apis/ote/extension/utils/runner.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import warnings
 
 import mmcv
+import mmdet
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/extension/utils/runner.py]: Is it used somehow?

File: mmdet/apis/ote/extension/utils/hooks.py [hunk 3]
```diff
@@ -487,3 +487,13 @@ def before_run(self, runner):
...
+
+
+@HOOKS.register_module()
+class StopLossNanTrainingHook(Hook):
+
+    def after_train_iter(self, runner):
+        if isnan(runner.outputs['loss'].item()):
+            logger.warning(f"Early Stopping since loss is NaN")
+            runner.should_stop = True
+            runner.should_validate = False
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/extension/utils/hooks.py]: Still of use?

File: mmdet/apis/ote/extension/utils/runner.py (hunk 2 above)
Comment (morkovka1337) [reply mmdet/apis/ote/extension/utils/runner.py]: Will remove it

File: mmdet/apis/ote/extension/utils/hooks.py (hunk 3 above)
Comment (morkovka1337) [reply mmdet/apis/ote/extension/utils/hooks.py]: Will remove

**Review by druzhkov-paul — APPROVED** [review]
